src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs

```diff
@@ -0,0 +1,157 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Holds on a <see cref="DocumentId"/> to <see cref="TextDocumentState"/> map and an ordering.
+    /// </summary>
+    internal readonly struct TextDocumentStates<TState>
+        where TState : TextDocumentState
+    {
+        public static readonly TextDocumentStates<TState> Empty = new(ImmutableArray<DocumentId>.Empty, ImmutableDictionary<DocumentId, TState>.Empty);
+
+        /// <summary>
+        /// Ordered <see cref="DocumentId"/>s. This is always an <see cref="ImmutableArray{T}"/>,
+        /// but we hold on its boxed instance to avoid re-boxing when returning this value from public APIs.
+        /// </summary>
+        public readonly IReadOnlyList<DocumentId> Ids;
+
+        public readonly ImmutableDictionary<DocumentId, TState> Map;
+
+        private TextDocumentStates(IReadOnlyList<DocumentId> ids, ImmutableDictionary<DocumentId, TState> map)
+        {
+            Debug.Assert(ids.All(id => map.ContainsKey(id)) && ids.Count == map.Count);
+
+            Ids = ids;
+            Map = map;
+        }
+
+        public TextDocumentStates(ImmutableArray<DocumentId> ids, ImmutableDictionary<DocumentId, TState> map)
+            : this((IReadOnlyList<DocumentId>)ids, map)
+        {
+        }
+
+        public TextDocumentStates(IReadOnlyList<TState> states)
+            : this(states.SelectAsArray(state => state.Id),
+                   states.ToImmutableDictionary(state => state.Id))
+        {
+        }
+
+        public TextDocumentStates(IEnumerable<DocumentInfo> infos, Func<DocumentInfo, TState> stateConstructor)
+            : this(infos.Select(info => info.Id).ToImmutableArray(),
+                   infos.ToImmutableDictionary(info => info.Id, stateConstructor))
+        {
+        }
+
+        public int Count
+            => Ids.Count;
+
+        public bool IsEmpty
+            => Count == 0;
+
+        public bool Contains(DocumentId id)
+            => Map.ContainsKey(id);
+
+        public TState? GetValue(DocumentId documentId)
+            => Map.TryGetValue(documentId, out var state) ? state : null;
+
+        public TState GetRequiredValue(DocumentId documentId)
+            => Map.TryGetValue(documentId, out var state) ? state : throw ExceptionUtilities.Unreachable;
+
+        public IEnumerable<TState> Values
+        {
+            get
+            {
+                foreach (var id in Ids)
+                {
+                    yield return Map[id];
```

Comment:
😨 This is scary (iterator allocation, plus every step requires a new lookup in the map).

Reply:
Would `SortedImmutableDictionary<DocumentId, (int ordinal, TState state)>` be better?

Reply:
I'll review where we enumerate all document states and see if they actually require ordering. If not they can enumerate the Map directly.

Reply:
Do you have a better data structure on mind that we can use to hold on the id state mapping that preserves order?

Reply:
Based on https://github.com/dotnet/roslyn/issues/48682 I'll do more optimizations here. 
Will try using SegmentedImmutableDictionary: https://github.com/dotnet/roslyn/pull/50189.